### PR TITLE
fix: session user can become null leading to unusable session

### PR DESCRIPTION
**Pull request by mswertz**
For elevated privileges clearActiveUser was called in a non-threadsafe, setting user==null for other server threads running in parallel. This is solved by wrapping these calls in a tx, which will give it a private database instance instead of using the shared database instance. In addition, added logic to set user=anonymous in case it still happens so user can log in again

This is expected to solve or reduce the issues described in issue #357 .

### CODE REVIEW

**Comment by mswertz** [comment]
[Bugfix for session becomes null](https://trello.com/c/PryVEvie/200-bugfix-for-session-becomes-null)

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java [hunk 1]
```diff
@@ -229,4 +232,26 @@ private static SqlSchemaMetadata renameTableTransaction(
     sm.tables.put(newName, tm);
     return sm;
   }
+
+  public List<String> getIneritedRolesForUser(String user) {
+    if (user == null) return new ArrayList<>();
+    // add cache because this function is called often
+    if (rolesCache != null) return rolesCache;
```
Comment (connoratrug) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java:239]: exposed private / state-full value of roles list , suggest te return immutable copy

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisSessionManager.java [hunk 2]
```diff
@@ -41,7 +41,20 @@ public MolgenisSession getSession(Request request) {
...
     MolgenisSession session = sessions.get(request.session().id());
-    logger.info("get session for user({})", session.getSessionUser());
+    if (session.getSessionUser() == null) {
+      logger.error(
+          "got user is NULL for session key ({}). Setting user=anonymous.",
+          session.getSessionUser(),
+          request.session().id());
+      // should never happen, but in previous bug user could be null
+      // therefore make anonymous so session is valid and user can log in again
+      session.getDatabase().setActiveUser("anonymous");
```
Comment (connoratrug) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisSessionManager.java:51]: swallows the error , why not just fail the request , as noted in the comment should not happen , could just be a bad request

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlDatabase.java [hunk 3]
```diff
@@ -315,10 +314,14 @@ public void grantCreateSchema(String user) {
   @Override
   public void setActiveUser(String username) {
     if (inTx) {
-      try {
-        jooq.execute("SET SESSION AUTHORIZATION {0}", name(MG_USER_PREFIX + username));
-      } catch (DataAccessException dae) {
-        throw new SqlMolgenisException("Set active user failed", dae);
+      if (username == null) {
+        this.clearActiveUser();
```
Comment (connoratrug) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlDatabase.java:318]: why would we allow to run clearUser outside a transaction ?

**Comment by connoratrug** [comment]
The PR seams to solve the race condition ( although no test is included to prove this ),  only kind of left with the feeling that these times of issues are the whole reason why Spring / AOP was invented.  The caller of the public database methods needs to keep thinking about state and race conditions then using the database methods, maybe look at wrapping this stuff into a smaller api that hide more impl details.  The mix of static source with statefull database objects and statefull transactions makes it pretty hard to follow whats going on imo ( resulting the these types of hard to reproduce / (prove) resolved ) kind of bugs.

**Review by connoratrug — CHANGES_REQUESTED** [review]
roles list should be immutable

**Comment by mswertz** [comment]
> The PR seams to solve the race condition ( although no test is included to prove this ), only kind of left with the feeling that these times of issues are the whole reason why Spring / AOP was invented. 

I actually don't think Spring would have shielded us from this problem. But instead we would have applied @Transactional

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlDatabase.java (hunk 3 above)
Comment (mswertz) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlDatabase.java:318]: > why would we allow to run clearUser outside a transaction ( see clearActiveUser impl)?

because the if(inTx) check already ensures we are in an transaction

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java (hunk 1 above)
Comment (mswertz) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java:239]: good point, will add before merging

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisSessionManager.java (hunk 2 above)
Comment (mswertz) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisSessionManager.java:51]: okay, will then throw an exception. At least we will find any issues sooner :-)

**Review by connoratrug — COMMENTED** [review]
see return copy comment

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java [hunk 4]
```diff
@@ -229,4 +233,26 @@ private static SqlSchemaMetadata renameTableTransaction(
...
+            tdb -> {
+              String current = tdb.getActiveUser();
+              tdb.clearActiveUser();
+              result.addAll(
+                  SqlSchemaMetadataExecutor.getInheritedRoleForUser(
+                      ((SqlDatabase) tdb).getJooq(), getName(), username));
+              tdb.setActiveUser(current);
+            });
+    rolesCache = result;
+    return result;
```
[the remark this answers is not in the document]
Comment (mswertz) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlSchemaMetadata.java:256]: ah, well spotted

**Review by connoratrug — APPROVED** [review]
